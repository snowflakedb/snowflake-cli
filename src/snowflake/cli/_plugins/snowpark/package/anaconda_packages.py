# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Dict, List, Set

from packaging.requirements import InvalidRequirement
from packaging.requirements import Requirement as PkgRequirement
from packaging.version import InvalidVersion, parse
from snowflake.cli._plugins.snowpark.models import Requirement
from snowflake.cli.api.exceptions import SnowflakeSQLExecutionError
from snowflake.cli.api.secure_path import SecurePath
from snowflake.cli.api.sql_execution import SqlExecutionMixin
from snowflake.connector import DictCursor

log = logging.getLogger(__name__)


@dataclass
class FilterRequirementsResult:
    """A dataclass to hold the results of parsing requirements files and dividing them into
    snowflake-supported vs other packages.
    """

    in_snowflake: List[Requirement]
    unavailable: List[Requirement]


@dataclass
class AvailablePackage:
    snowflake_name: str
    versions: Set[str]

    def iter_versions(self):
        for version in self.versions:
            yield parse(version)

    def is_required_version_available(self, requirement: Requirement) -> bool:
        try:
            package_specifiers = PkgRequirement(requirement.line).specifier
            return any(
                version in package_specifiers for version in self.iter_versions()
            )
        except (InvalidVersion, InvalidRequirement):
            # fail-safe for non-pep508 formats
            return False


class AnacondaPackages:
    def __init__(self, packages: Dict[str, AvailablePackage]):
        """
        [packages] should be a dictionary mapping package name to AnacondaPackageData object.
        All package names should be provided in wheel escape format:
        https://peps.python.org/pep-0491/#escaping-and-unicode
        """
        self._packages = packages

    @classmethod
    def empty(cls):
        return cls({})

    def is_package_available(
        self, package: Requirement, skip_version_check: bool = False
    ) -> bool:
        """
        Checks of a requirement is available in the Snowflake Anaconda Channel.

        As Snowflake currently doesn't support extra syntax (ex. `jinja2[diagrams]`), if such
        extra is present in the dependency, we mark it as unavailable.
        """
        if not package.name or package.extras:
            return False
        if package.name not in self._packages:
            return False
        if skip_version_check or not package.specs:
            return True
        if any(spec[0] == "!=" for spec in package.specs):
            # Snowflake doesn't support '!=' so we need to resolve this requirement externally
            return False
        return self._packages[package.name].is_required_version_available(package)

    def package_latest_version(self, package: Requirement) -> str | None:
        """Returns the latest version of the package or None if the latest version can't be determined."""
        if package.name not in self._packages:
            return None
        try:
            return str(max(self._packages[package.name].iter_versions()))
        except InvalidVersion:
            # fail-safe for non-pep8 versions
            return None

    def package_versions(self, package: Requirement) -> List[str]:
        """Returns list of available versions of the package."""
        if package.name not in self._packages:
            return []
        package_data = self._packages[package.name]
        try:
            return list(
                str(x) for x in sorted(package_data.iter_versions(), reverse=True)
            )
        except InvalidVersion:
            return list(sorted(package_data.versions, reverse=True))

    def filter_available_packages(
        self, packages: List[Requirement], skip_version_check: bool = False
    ) -> FilterRequirementsResult:
        """
        Checks if a list of packages are available in the Snowflake Anaconda channel.
        Returns an object with two attributes: 'snowflake' and 'other'.
        Each key contains a list of Requirement object.

        Parameters:
            packages (List[Requirement]) - list of requirements to be checked
            skip_version_check (bool) - skip comparing versions of packages

        Returns:
            result (FilterRequirementsResult) - object containing two arguments:
              - in_snowflake - packages available in conda
              - unavailable - packages not available in conda
        """
        result = FilterRequirementsResult([], [])
        for package in packages:
            if self.is_package_available(
                package, skip_version_check=skip_version_check
            ):
                result.in_snowflake.append(package)
            else:
                log.info(
                    "'%s' not found in Snowflake Anaconda channel (or ignored)...",
                    package.name,
                )
                result.unavailable.append(package)
        return result

    def write_requirements_file_in_snowflake_format(
        self,
        file_path: SecurePath,
        requirements: List[Requirement],
    ):
        """Saves requirements to a file in format accepted by Snowflake SQL commands."""
        log.info("Writing requirements into file %s", file_path.path)
        formatted_requirements = []
        for requirement in requirements:
            if requirement.name and requirement.name in self._packages:
                snowflake_name = self._packages[requirement.name].snowflake_name
                formatted_requirements.append(
                    snowflake_name + requirement.formatted_specs
                )

        if formatted_requirements:
            file_path.write_text("\n".join(formatted_requirements))


class AnacondaPackagesManager(SqlExecutionMixin):
    _snowflake_channel_url: str = (
        "https://repo.anaconda.com/pkgs/snowflake/channeldata.json"
    )

    def find_packages_available_in_snowflake_anaconda(self) -> AnacondaPackages:
        """
        Finds Python packages available in Snowflake to use in functions and stored procedures.
        It tries to get the list of packages using SQL query
        but if the try fails then the fallback is to parse JSON containing info about Snowflake's Anaconda channel.
        """
        packages = self._query_snowflake_for_available_packages()
        return AnacondaPackages(packages)

    def _query_snowflake_for_available_packages(self) -> dict[str, AvailablePackage]:
        cursor = self.execute_query(
            "select package_name, version from snowflake.information_schema.packages where language = 'python'",
            cursor_class=DictCursor,
        )
        if cursor.rowcount is None or cursor.rowcount == 0:
            raise SnowflakeSQLExecutionError()
        packages: dict[str, AvailablePackage] = {}
        for row in cursor:
            if not (package_name := row["PACKAGE_NAME"]):
                continue
            if not (version := row["VERSION"]):
                continue
            standardized_name = Requirement.standardize_name(package_name)
            if standardized_name in packages:
                packages[standardized_name].versions.add(version)
            else:
                packages[standardized_name] = AvailablePackage(
                    snowflake_name=package_name, versions={version}
                )
        return packages
