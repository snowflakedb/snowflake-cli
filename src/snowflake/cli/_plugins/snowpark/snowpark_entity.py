from enum import Enum
from pathlib import Path
from typing import Generic, List, Optional, TypeVar

from click import ClickException
from snowflake.cli._plugins.nativeapp.feature_flags import FeatureFlag
from snowflake.cli._plugins.snowpark import package_utils
from snowflake.cli._plugins.snowpark.common import (
    DEFAULT_RUNTIME,
    map_path_mapping_to_artifact,
    zip_and_copy_artifacts_to_deploy,
)
from snowflake.cli._plugins.snowpark.package.anaconda_packages import (
    AnacondaPackages,
    AnacondaPackagesManager,
)
from snowflake.cli._plugins.snowpark.package_utils import (
    DownloadUnavailablePackagesResult,
)
from snowflake.cli._plugins.snowpark.snowpark_entity_model import (
    FunctionEntityModel,
    ProcedureEntityModel,
)
from snowflake.cli._plugins.snowpark.snowpark_project_paths import SnowparkProjectPaths
from snowflake.cli._plugins.snowpark.zipper import zip_dir
from snowflake.cli._plugins.workspace.context import ActionContext
from snowflake.cli.api.entities.common import EntityBase
from snowflake.cli.api.secure_path import SecurePath
from snowflake.connector import ProgrammingError

T = TypeVar("T")


class CreateMode(
    str, Enum
):  # This should probably be moved to some common place, think where
    create = "CREATE"
    create_or_replace = "CREATE OR REPLACE"
    create_if_not_exists = "CREATE IF NOT EXISTS"


class SnowparkEntity(EntityBase[Generic[T]]):
    def __init__(self, *args, **kwargs):

        if not FeatureFlag.ENABLE_NATIVE_APP_CHILDREN.is_enabled():
            raise NotImplementedError("Snowpark entity is not implemented yet")
        super().__init__(*args, **kwargs)

    def action_bundle(
        self,
        action_ctx: ActionContext,
        ignore_anaconda: bool,
        skip_version_check: bool,
        output_dir: Path | None = None,
        index_url: str | None = None,
        allow_shared_libraries: bool = False,
        *args,
        **kwargs,
    ) -> List[Path]:
        return self.bundle(
            ignore_anaconda,
            skip_version_check,
            output_dir,
            index_url,
            allow_shared_libraries,
        )

    def action_deploy(
        self,
        action_ctx: ActionContext,
        mode: CreateMode = CreateMode.create,
        *args,
        **kwargs,
    ):
        # TODO: After introducing bundle map, we should introduce file copying part here
        return self.deploy(mode, *args, **kwargs)

    def action_drop(self, action_ctx: ActionContext, *args, **kwargs):
        return self._execute_query(self.get_drop_sql())

    def action_describe(self, action_ctx: ActionContext, *args, **kwargs):
        return self._execute_query(self.get_describe_sql())

    def action_execute(
        self,
        action_ctx: ActionContext,
        execution_arguments: List[str] | None = None,
        *args,
        **kwargs,
    ):
        return self._execute_query(self.get_execute_sql(execution_arguments))

    def bundle(
        self,
        ignore_anaconda: bool,
        skip_version_check: bool,
        output_dir: Path | None = None,
        index_url: str | None = None,
        allow_shared_libraries: bool = False,
    ) -> List[Path]:
        """
        Bundles the entity artifacts and dependencies into a directory.
        Parameters:
            output_dir: The directory to output the bundled artifacts to. Defaults to output dir in project root
            ignore_anaconda: If True, ignores anaconda check and tries to download all packages using pip
            skip_version_check: If True, skips version check when downloading packages
            index_url: The index URL to use when downloading packages, if none set - default pip index is used (in most cases- Pypi)
            allow_shared_libraries: If not set to True, using dependency with .so/.dll files will raise an exception
        Returns:
        """
        # 0 Create a directory for the entity
        project_paths = SnowparkProjectPaths(
            project_root=self.root.absolute(),
        )
        if not output_dir:
            output_dir = project_paths.bundle_root
        if not output_dir.exists():  # type: ignore[union-attr]
            SecurePath(output_dir).mkdir(parents=True)

        # 1 Check if requirements exits
        if (self.root / "requirements.txt").exists():
            download_results = self._process_requirements(
                bundle_dir=output_dir,  # type: ignore
                archive_name="dependencies.zip",
                requirements_file=SecurePath(self.root / "requirements.txt"),
                ignore_anaconda=ignore_anaconda,
                skip_version_check=skip_version_check,
                index_url=index_url,
                allow_shared_libraries=allow_shared_libraries,
            )

        # 2 get the artifacts list
        artifacts = map_path_mapping_to_artifact(project_paths, self.model.artifacts)
        from snowflake.cli.api.feature_flags import FeatureFlag

        if FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_enabled():
            return zip_and_copy_artifacts_to_deploy(
                artifacts, project_paths.bundle_root
            )
        else:
            copied_files = []
            for artefact in artifacts:
                artefact.build()
                copied_files.append(artefact.post_build_path)
            return copied_files

    def check_if_exists(
        self, action_ctx: ActionContext
    ) -> bool:  # TODO it should return current state, so we know if update is necessary
        try:
            current_state = self.action_describe(action_ctx)
            return True
        except ProgrammingError:
            return False

    def deploy(self, mode: CreateMode = CreateMode.create, *args, **kwargs):
        return self._execute_query(self.get_deploy_sql(mode))

    def get_deploy_sql(self, mode: CreateMode):
        query = [
            f"{mode.value} {self.model.type.upper()} {self.identifier}",
            "COPY GRANTS",
            f"RETURNS {self.model.returns}",
            f"LANGUAGE PYTHON",
            f"RUNTIME_VERSION '{self.model.runtime or DEFAULT_RUNTIME}'",
            f"IMPORTS={','.join(self.model.imports)}",  # TODO: Add source files here after introducing bundlemap
            f"HANDLER='{self.model.handler}'",
        ]

        if self.model.external_access_integrations:
            query.append(self.model.get_external_access_integrations_sql())

        if self.model.secrets:
            query.append(self.model.get_secrets_sql())

        if self.model.type == "procedure" and self.model.execute_as_caller:
            query.append("EXECUTE AS CALLER")

        return "\n".join(query)

    def get_execute_sql(self, execution_arguments: List[str] | None = None):
        raise NotImplementedError

    def _process_requirements(  # TODO: maybe leave all the logic with requirements here - so download, write requirements file etc.
        self,
        bundle_dir: Path,
        archive_name: str,  # TODO: not the best name, think of something else
        requirements_file: Optional[SecurePath],
        ignore_anaconda: bool,
        skip_version_check: bool = False,
        index_url: Optional[str] = None,
        allow_shared_libraries: bool = False,
    ) -> DownloadUnavailablePackagesResult:
        """
        Processes the requirements file and downloads the dependencies
        Parameters:

        """
        anaconda_packages_manager = AnacondaPackagesManager()
        with SecurePath.temporary_directory() as tmp_dir:
            requirements = package_utils.parse_requirements(requirements_file)
            anaconda_packages = (
                AnacondaPackages.empty()
                if ignore_anaconda
                else anaconda_packages_manager.find_packages_available_in_snowflake_anaconda()
            )
            download_result = package_utils.download_unavailable_packages(
                requirements=requirements,
                target_dir=tmp_dir,
                anaconda_packages=anaconda_packages,
                skip_version_check=skip_version_check,
                pip_index_url=index_url,
            )

            if download_result.anaconda_packages:
                anaconda_packages.write_requirements_file_in_snowflake_format(
                    file_path=SecurePath(bundle_dir / "requirements.txt"),
                    requirements=download_result.anaconda_packages,
                )

            if download_result.downloaded_packages_details:
                if (
                    package_utils.detect_and_log_shared_libraries(
                        download_result.downloaded_packages_details
                    )
                    and not allow_shared_libraries
                ):
                    raise ClickException(
                        "Some packages contain shared (.so/.dll) libraries. "
                        "Try again with allow_shared_libraries_flag."
                    )

                zip_dir(
                    source=tmp_dir,
                    dest_zip=bundle_dir / archive_name,
                )

        return download_result


class FunctionEntity(SnowparkEntity[FunctionEntityModel]):
    """
    A single UDF
    """

    # TO THINK OF
    # Where will we get imports? Should we rely on bundle map? Or should it be self-sufficient in this matter?

    def get_execute_sql(
        self, execution_arguments: List[str] | None = None, *args, **kwargs
    ):
        if not execution_arguments:
            execution_arguments = []
        return (
            f"SELECT {self.fqn}({', '.join([str(arg) for arg in execution_arguments])})"
        )


class ProcedureEntity(SnowparkEntity[ProcedureEntityModel]):
    """
    A stored procedure
    """

    def get_execute_sql(
        self,
        execution_arguments: List[str] | None = None,
    ):
        if not execution_arguments:
            execution_arguments = []
        return (
            f"CALL {self.fqn}({', '.join([str(arg) for arg in execution_arguments])})"
        )
