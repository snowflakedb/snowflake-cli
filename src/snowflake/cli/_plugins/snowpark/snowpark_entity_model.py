# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import glob
from typing import List, Literal, Optional, Union

from pydantic import Field, field_validator
from snowflake.cli.api.feature_flags import FeatureFlag
from snowflake.cli.api.identifiers import FQN
from snowflake.cli.api.project.schemas.entities.common import (
    EntityModelBaseWithArtifacts,
    ExternalAccessBaseModel,
    ImportsBaseModel,
    PathMapping,
)
from snowflake.cli.api.project.schemas.updatable_model import (
    DiscriminatorField,
)
from snowflake.cli.api.project.schemas.v1.snowpark.argument import Argument


class SnowparkEntityModel(
    EntityModelBaseWithArtifacts, ExternalAccessBaseModel, ImportsBaseModel
):
    handler: str = Field(
        title="Function’s or procedure’s implementation of the object inside source module",
        examples=["functions.hello_function"],
    )
    returns: str = Field(
        title="Type of the result"
    )  # TODO: again, consider Literal/Enum
    signature: Union[str, List[Argument]] = Field(
        title="The signature parameter describes consecutive arguments passed to the object"
    )
    runtime: Optional[Union[str, float]] = Field(
        title="Python version to use when executing ", default=None
    )
    stage: str = Field(title="Stage in which artifacts will be stored")

    @field_validator("artifacts")
    @classmethod
    def _convert_artifacts(cls, artifacts: Union[dict, str]):
        _artifacts = []
        for artifact in artifacts:
            if (
                (isinstance(artifact, str) and glob.has_magic(artifact))
                or (isinstance(artifact, PathMapping) and glob.has_magic(artifact.src))
            ) and FeatureFlag.ENABLE_SNOWPARK_GLOB_SUPPORT.is_disabled():
                raise ValueError(
                    "If you want to use glob patterns in artifacts, you need to enable the Snowpark new build feature flag (enable_snowpark_glob_support=true)"
                )
            if isinstance(artifact, PathMapping):
                _artifacts.append(artifact)
            else:
                _artifacts.append(PathMapping(src=artifact))
        return _artifacts

    @field_validator("runtime")
    @classmethod
    def convert_runtime(cls, runtime_input: Union[str, float]) -> str:
        if isinstance(runtime_input, float):
            return str(runtime_input)
        return runtime_input

    @property
    def udf_sproc_identifier(self) -> UdfSprocIdentifier:
        return UdfSprocIdentifier.from_definition(self)


class ProcedureEntityModel(SnowparkEntityModel):
    type: Literal["procedure"] = DiscriminatorField()  # noqa: A003
    execute_as_caller: Optional[bool] = Field(
        title="Determine whether the procedure is executed with the privileges of "
        "the owner (you) or with the privileges of the caller",
        default=False,
    )


class FunctionEntityModel(SnowparkEntityModel):
    type: Literal["function"] = DiscriminatorField()  # noqa: A003


class UdfSprocIdentifier:
    def __init__(self, identifier: FQN, arg_names, arg_types, arg_defaults):
        self._identifier = identifier
        self._arg_names = arg_names
        self._arg_types = arg_types
        self._arg_defaults = arg_defaults

    def _identifier_from_signature(self, sig: List[str], for_sql: bool = False):
        signature = self._comma_join(sig)
        id_ = self._identifier.sql_identifier if for_sql else self._identifier
        return f"{id_}({signature})"

    @staticmethod
    def _comma_join(*args):
        return ", ".join(*args)

    @property
    def identifier_with_arg_names(self):
        return self._identifier_from_signature(self._arg_names)

    @property
    def identifier_with_arg_types(self):
        return self._identifier_from_signature(self._arg_types)

    @property
    def identifier_with_arg_names_types(self):
        sig = [f"{n} {t}" for n, t in zip(self._arg_names, self._arg_types)]
        return self._identifier_from_signature(sig)

    @property
    def identifier_with_arg_names_types_defaults(self):
        return self._identifier_from_signature(self._full_signature())

    def _is_signature_type_a_string(self, sig_type: str) -> bool:
        return sig_type.lower() in ["string", "varchar"]

    def _full_signature(self):
        sig = []
        for name, _type, _default in zip(
            self._arg_names, self._arg_types, self._arg_defaults
        ):
            s = f"{name} {_type}"
            if _default is not None:
                if (
                    self._is_signature_type_a_string(_type)
                    and _default.lower() != "null"
                ):
                    _default = f"'{_default}'"
                s += f" default {_default}"
            sig.append(s)
        return sig

    @property
    def identifier_for_sql(self):
        return self._identifier_from_signature(self._full_signature(), for_sql=True)

    @classmethod
    def from_definition(cls, udf_sproc: SnowparkEntityModel):
        names = []
        types = []
        defaults = []
        if udf_sproc.signature and udf_sproc.signature != "null":
            for arg in udf_sproc.signature:
                names.append(arg.name)  # type:ignore
                types.append(arg.arg_type)  # type:ignore
                defaults.append(arg.default)  # type:ignore

        identifier = udf_sproc.fqn.using_context()
        return cls(identifier, names, types, defaults)
