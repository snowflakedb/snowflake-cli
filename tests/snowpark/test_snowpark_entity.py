from pathlib import Path
from unittest import mock

import pytest
import yaml
from snowflake.cli._plugins.snowpark.package.anaconda_packages import (
    AnacondaPackages,
    AvailablePackage,
)
from snowflake.cli._plugins.snowpark.snowpark_entity import (
    CreateMode,
    FunctionEntity,
    ProcedureEntity,
)
from snowflake.cli._plugins.snowpark.snowpark_entity_model import FunctionEntityModel
from snowflake.cli._plugins.workspace.context import ActionContext, WorkspaceContext

from tests.testing_utils.mock_config import mock_config_key

CONNECTOR = "snowflake.connector.connect"
CONTEXT = ""
EXECUTE_QUERY = "snowflake.cli.api.sql_execution.BaseSqlExecutor.execute_query"
ANACONDA_PACKAGES = "snowflake.cli._plugins.snowpark.package.anaconda_packages.AnacondaPackagesManager.find_packages_available_in_snowflake_anaconda"


@pytest.fixture
def example_function_workspace(
    project_directory,
):  # TODO: try to make a common fixture for all entities
    with mock_config_key("enable_native_app_children", True):
        with project_directory("snowpark_functions_v2") as pdir:
            with Path(pdir / "snowflake.yml").open() as definition_file:
                definition = yaml.safe_load(definition_file)
                model = FunctionEntityModel(
                    **definition.get("entities", {}).get("func1")
                )

                workspace_context = WorkspaceContext(
                    console=mock.MagicMock(),
                    project_root=pdir,
                    get_default_role=lambda: "test_role",
                    get_default_warehouse=lambda: "test_warehouse",
                )

                return (
                    FunctionEntity(workspace_ctx=workspace_context, entity_model=model),
                    ActionContext(
                        get_entity=lambda *args: None,
                    ),
                )


def test_cannot_instantiate_without_feature_flag():
    with pytest.raises(NotImplementedError) as err:
        FunctionEntity()
    assert str(err.value) == "Snowpark entity is not implemented yet"

    with pytest.raises(NotImplementedError) as err:
        ProcedureEntity()
    assert str(err.value) == "Snowpark entity is not implemented yet"


@mock.patch(ANACONDA_PACKAGES)
def test_nativeapp_children_interface_old_build(
    mock_anaconda, example_function_workspace, snapshot
):
    mock_anaconda.return_value = AnacondaPackages(
        {
            "pandas": AvailablePackage("pandas", "1.2.3"),
            "numpy": AvailablePackage("numpy", "1.2.3"),
            "snowflake_snowpark_python": AvailablePackage(
                "snowflake_snowpark_python", "1.2.3"
            ),
        }
    )

    sl, action_context = example_function_workspace

    sl.bundle(None, False, False, None, False)

    deploy_sql_str = sl.get_deploy_sql(CreateMode.create)
    grant_sql_str = sl.get_usage_grant_sql(app_role="app_role")

    assert deploy_sql_str == snapshot
    assert (
        grant_sql_str
        == f"GRANT USAGE ON FUNCTION IDENTIFIER('func1') TO ROLE app_role;"
    )


@mock.patch(ANACONDA_PACKAGES)
def test_nativeapp_children_interface(
    mock_anaconda,
    example_function_workspace,
    snapshot,
    enable_snowpark_glob_support_feature_flag,
):
    mock_anaconda.return_value = AnacondaPackages(
        {
            "pandas": AvailablePackage("pandas", "1.2.3"),
            "numpy": AvailablePackage("numpy", "1.2.3"),
            "snowflake_snowpark_python": AvailablePackage(
                "snowflake_snowpark_python", "1.2.3"
            ),
        }
    )

    sl, action_context = example_function_workspace

    sl.bundle(None, False, False, None, False)
    bundle_artifact = (
        sl.root / "output" / "bundle" / "snowpark" / "my_snowpark_project" / "app.py"
    )
    deploy_sql_str = sl.get_deploy_sql(CreateMode.create)
    grant_sql_str = sl.get_usage_grant_sql(app_role="app_role")

    assert bundle_artifact.exists()
    assert deploy_sql_str == snapshot
    assert (
        grant_sql_str
        == f"GRANT USAGE ON FUNCTION IDENTIFIER('func1') TO ROLE app_role;"
    )


@mock.patch(EXECUTE_QUERY)
def test_action_describe(mock_execute, example_function_workspace):
    entity, action_context = example_function_workspace
    result = entity.action_describe(action_context)

    mock_execute.assert_called_with("DESCRIBE FUNCTION IDENTIFIER('func1');")


@mock.patch(EXECUTE_QUERY)
def test_action_drop(mock_execute, example_function_workspace):
    entity, action_context = example_function_workspace
    result = entity.action_drop(action_context)

    mock_execute.assert_called_with("DROP FUNCTION IDENTIFIER('func1');")


@pytest.mark.parametrize(
    "execution_arguments", [None, ["arg1", "arg2"], ["foo", 42, "bar"]]
)
@mock.patch(EXECUTE_QUERY)
def test_action_execute(
    mock_execute, execution_arguments, example_function_workspace, snapshot
):
    entity, action_context = example_function_workspace
    result = entity.action_execute(action_context, execution_arguments)

    mock_execute.assert_called_with(snapshot)


@mock.patch(ANACONDA_PACKAGES)
def test_bundle_old_build(mock_anaconda, example_function_workspace):
    mock_anaconda.return_value = AnacondaPackages(
        {
            "pandas": AvailablePackage("pandas", "1.2.3"),
            "numpy": AvailablePackage("numpy", "1.2.3"),
            "snowflake_snowpark_python": AvailablePackage(
                "snowflake_snowpark_python", "1.2.3"
            ),
        }
    )
    entity, action_context = example_function_workspace
    entity.action_bundle(action_context, None, False, False, None, False)

    assert (entity.root / "app.py").exists()


@mock.patch(ANACONDA_PACKAGES)
def test_bundle(
    mock_anaconda, example_function_workspace, enable_snowpark_glob_support_feature_flag
):
    mock_anaconda.return_value = AnacondaPackages(
        {
            "pandas": AvailablePackage("pandas", "1.2.3"),
            "numpy": AvailablePackage("numpy", "1.2.3"),
            "snowflake_snowpark_python": AvailablePackage(
                "snowflake_snowpark_python", "1.2.3"
            ),
        }
    )
    entity, action_context = example_function_workspace
    entity.action_bundle(action_context, None, False, False, None, False)

    output = entity.root / "output" / "bundle" / "snowpark"  # noqa
    assert output.exists()
    assert (output / "my_snowpark_project" / "app.py").exists()


def test_describe_function_sql(example_function_workspace):
    entity, _ = example_function_workspace
    assert entity.get_describe_sql() == "DESCRIBE FUNCTION IDENTIFIER('func1');"


def test_drop_function_sql(example_function_workspace):
    entity, _ = example_function_workspace
    assert entity.get_drop_sql() == "DROP FUNCTION IDENTIFIER('func1');"


@pytest.mark.parametrize(
    "execution_arguments", [None, ["arg1", "arg2"], ["foo", 42, "bar"]]
)
def test_function_get_execute_sql(
    execution_arguments, example_function_workspace, snapshot
):
    entity, _ = example_function_workspace
    assert entity.get_execute_sql(execution_arguments) == snapshot


@pytest.mark.parametrize(
    "mode",
    [CreateMode.create, CreateMode.create_or_replace, CreateMode.create_if_not_exists],
)
def test_get_deploy_sql(mode, example_function_workspace, snapshot):
    entity, _ = example_function_workspace
    assert entity.get_deploy_sql(mode) == snapshot


def test_get_usage_grant_sql(example_function_workspace):
    entity, _ = example_function_workspace
    assert (
        entity.get_usage_grant_sql("test_role")
        == "GRANT USAGE ON FUNCTION IDENTIFIER('func1') TO ROLE test_role;"
    )
