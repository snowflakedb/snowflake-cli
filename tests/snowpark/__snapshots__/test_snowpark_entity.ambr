# serializer version: 1
# name: test_action_execute[None]
  'SELECT func1()'
# ---
# name: test_action_execute[execution_arguments1]
  'SELECT func1(arg1, arg2)'
# ---
# name: test_action_execute[execution_arguments2]
  'SELECT func1(foo, 42, bar)'
# ---
# name: test_function_get_execute_sql[None]
  'SELECT func1()'
# ---
# name: test_function_get_execute_sql[execution_arguments1]
  'SELECT func1(arg1, arg2)'
# ---
# name: test_function_get_execute_sql[execution_arguments2]
  'SELECT func1(foo, 42, bar)'
# ---
# name: test_get_deploy_sql[CREATE IF NOT EXISTS]
  '''
  CREATE IF NOT EXISTS FUNCTION IDENTIFIER('func1')
  COPY GRANTS
  RETURNS string
  LANGUAGE PYTHON
  RUNTIME_VERSION '3.10'
  IMPORTS=
  HANDLER='app.func1_handler'
  '''
# ---
# name: test_get_deploy_sql[CREATE OR REPLACE]
  '''
  CREATE OR REPLACE FUNCTION IDENTIFIER('func1')
  COPY GRANTS
  RETURNS string
  LANGUAGE PYTHON
  RUNTIME_VERSION '3.10'
  IMPORTS=
  HANDLER='app.func1_handler'
  '''
# ---
# name: test_get_deploy_sql[CREATE]
  '''
  CREATE FUNCTION IDENTIFIER('func1')
  COPY GRANTS
  RETURNS string
  LANGUAGE PYTHON
  RUNTIME_VERSION '3.10'
  IMPORTS=
  HANDLER='app.func1_handler'
  '''
# ---
# name: test_nativeapp_children_interface
  '''
  CREATE FUNCTION IDENTIFIER('func1')
  COPY GRANTS
  RETURNS string
  LANGUAGE PYTHON
  RUNTIME_VERSION '3.10'
  IMPORTS=
  HANDLER='app.func1_handler'
  '''
# ---
# name: test_nativeapp_children_interface_old_build
  '''
  CREATE FUNCTION IDENTIFIER('func1')
  COPY GRANTS
  RETURNS string
  LANGUAGE PYTHON
  RUNTIME_VERSION '3.10'
  IMPORTS=
  HANDLER='app.func1_handler'
  '''
# ---
